src/common/droid-util.c

```diff
@@ -1906,48 +1910,105 @@ static pa_droid_stream *droid_stream_new(pa_droid_hw_module *module) {
     return s;
 }
 
+static bool stream_config_fill(audio_devices_t devices,
+                               pa_sample_spec *sample_spec,
+                               pa_channel_map *channel_map,
+                               struct audio_config *config) {
+    audio_format_t hal_audio_format = 0;
+    audio_channel_mask_t hal_channel_mask = 0;
+    bool voicecall_record = false;
+
+    pa_assert(sample_spec);
+    pa_assert(channel_map);
+    pa_assert(config);
+
+#if AUDIO_API_VERSION_MAJ >= 2
+    if ((devices & ~AUDIO_DEVICE_BIT_IN) & AUDIO_DEVICE_IN_VOICE_CALL)
+#else
+    if (devices & AUDIO_DEVICE_IN_VOICE_CALL)
+#endif
+        voicecall_record = true;
+
+    if (!pa_convert_format(sample_spec->format, CONV_FROM_PA, &hal_audio_format)) {
+        pa_log("Sample spec format %u not supported.", sample_spec->format);
+        goto fail;
+    }
+
+    if (devices & AUDIO_DEVICE_IN_ALL) {
+        for (int i = 0; i < channel_map->channels; i++) {
+            audio_channel_mask_t c;
+            if (!pa_convert_input_channel(channel_map->map[i], CONV_FROM_PA, &c)) {
+                pa_log("Failed to convert channel map.");
+                goto fail;
+            }
+            hal_channel_mask |= c;
+        }
+    } else {
+        for (int i = 0; i < channel_map->channels; i++) {
+            audio_channel_mask_t c;
+            if (!pa_convert_output_channel(channel_map->map[i], CONV_FROM_PA, &c)) {
+                pa_log("Failed to convert channel map.");
+                goto fail;
+            }
+            hal_channel_mask |= c;
+        }
+    }
+
+    if (voicecall_record) {
+        pa_channel_map_init_mono(channel_map);
+        sample_spec->channels = 1;
+        /* Only allow recording both downlink and uplink. */
+#if defined(QCOM_HARDWARE)
+  #if (ANDROID_VERSION_MAJOR <= 4) && defined(HAVE_ENUM_AUDIO_CHANNEL_IN_VOICE_CALL_MONO)
+        hal_channel_mask = AUDIO_CHANNEL_IN_VOICE_CALL_MONO;
+  #else
+        hal_channel_mask = AUDIO_CHANNEL_IN_MONO;
+  #endif
+#else
+        hal_channel_mask = AUDIO_CHANNEL_IN_VOICE_UPLINK | AUDIO_CHANNEL_IN_VOICE_DNLINK;
+#endif
+    }
+
+    memset(config, 0, sizeof(struct audio_config));
```

Comment:
Matter of taste, but: "memset(config, 0, sizeof *config);" == I do not need to check if config is actually of type "struct audio_config *".